// vuestic-ui/dist/esm/vuestic-components2.js
import { VaAccordion } from './index.js';
import { VaAffix } from './index2.js';
import { VaAlert } from './index3.js';
import { VaAppBar } from './index4.js';
import { VaAvatar } from './index5.js';
import { VaBacktop } from './index6.js';
import { VaBadge } from './index7.js';
import { VaBreadcrumbs, VaBreadcrumbsItem } from './index8.js';
import { VaButtonDropdown } from './index9.js';
import { VaButtonGroup } from './index10.js';
import { VaButtonToggle } from './index11.js';
import { VaButton } from './index12.js';
import { VaCard, VaCardActions, VaCardContent, VaCardTitle } from './index13.js';
import { VaCarousel } from './index14.js';
import { VaCheckbox } from './index15.js';
import { VaChip } from './index16.js';
import { VaCollapse } from './index17.js';
import { VaColorIndicator } from './index18.js';
import { VaColorInput } from './index19.js';
import { VaColorPalette } from './index20.js';
import { VaContent } from './index21.js';
import { VaCounter } from './index22.js';
import { VaDataTable } from './index23.js';
import { VaDateInput } from './index24.js';
import { VaDatePicker } from './index25.js';
import { VaDivider } from './index26.js';
import { VaDropdown, VaDropdownContent } from './index27.js';
import { VaFileUpload } from './index28.js';
import { VaForm } from './index29.js';
import { VaHover } from './index30.js';
import { VaIcon } from './index31.js';
import { VaImage } from './index32.js';
import { VaInfiniteScroll } from './index33.js';
import { VaInnerLoading } from './index34.js';
import { VaInput } from './index35.js';
import { VaList, VaListItem, VaListItemLabel, VaListItemSection, VaListLabel, VaListSeparator } from './index36.js';
import { VaNavbar, VaNavbarItem } from './index37.js';
import { VaOptionList } from './index38.js';
import { VaPagination } from './index39.js';
import { VaParallax } from './index40.js';
import { VaPopover } from './index41.js';
import { VaProgressBar } from './index42.js';
import { VaProgressCircle } from './index43.js';
import { VaRadio } from './index44.js';
import { VaRating } from './index45.js';
import { VaSelect } from './index46.js';
import { VaSidebar, VaSidebarItem, VaSidebarItemContent, VaSidebarItemTitle } from './index47.js';
import { VaSlider } from './index48.js';
import { VaSwitch } from './index49.js';
import { VaTabs, VaTab } from './index50.js';
import { VaTimeInput } from './index51.js';
import { VaTimePicker } from './index52.js';
import { VaTimeline, VaTimelineItem } from './index53.js';
import { VaTreeCategory, VaTreeNode, VaTreeView } from './index54.js';
import { V as VaConfig } from './VaConfig.js';
import { V as VaMessageList } from './index67.js';
import { V as VaModal } from './VaModal.js';
import { s as script } from './VaSeparator.vue_vue_type_script_lang.js';
import { s as script$1 } from './VaSpacer.vue_vue_type_script_lang.js';

var vuesticComponentMap = /*#__PURE__*/Object.freeze({
  __proto__: null,
  VaAccordion: VaAccordion,
  VaAffix: VaAffix,
  VaAlert: VaAlert,
  VaAppBar: VaAppBar,
  VaAvatar: VaAvatar,
  VaBacktop: VaBacktop,
  VaBadge: VaBadge,
  VaBreadcrumbs: VaBreadcrumbs,
  VaBreadcrumbsItem: VaBreadcrumbsItem,
  VaButton: VaButton,
  VaButtonDropdown: VaButtonDropdown,
  VaButtonGroup: VaButtonGroup,
  VaButtonToggle: VaButtonToggle,
  VaCard: VaCard,
  VaCardActions: VaCardActions,
  VaCardContent: VaCardContent,
  VaCardTitle: VaCardTitle,
  VaCarousel: VaCarousel,
  VaCheckbox: VaCheckbox,
  VaChip: VaChip,
  VaCollapse: VaCollapse,
  VaColorIndicator: VaColorIndicator,
  VaColorInput: VaColorInput,
  VaColorPalette: VaColorPalette,
  VaConfig: VaConfig,
  VaContent: VaContent,
  VaCounter: VaCounter,
  VaDataTable: VaDataTable,
  VaDateInput: VaDateInput,
  VaDatePicker: VaDatePicker,
  VaDivider: VaDivider,
  VaDropdown: VaDropdown,
  VaDropdownContent: VaDropdownContent,
  VaFileUpload: VaFileUpload,
  VaForm: VaForm,
  VaHover: VaHover,
  VaIcon: VaIcon,
  VaImage: VaImage,
  VaInfiniteScroll: VaInfiniteScroll,
  VaInnerLoading: VaInnerLoading,
  VaInput: VaInput,
  VaList: VaList,
  VaListItem: VaListItem,
  VaListItemLabel: VaListItemLabel,
  VaListItemSection: VaListItemSection,
  VaListLabel: VaListLabel,
  VaListSeparator: VaListSeparator,
  VaMessageList: VaMessageList,
  VaModal: VaModal,
  VaNavbar: VaNavbar,
  VaNavbarItem: VaNavbarItem,
  VaOptionList: VaOptionList,
  VaPagination: VaPagination,
  VaParallax: VaParallax,
  VaPopover: VaPopover,
  VaProgressBar: VaProgressBar,
  VaProgressCircle: VaProgressCircle,
  VaRadio: VaRadio,
  VaRating: VaRating,
  VaSelect: VaSelect,
  VaSeparator: script,
  VaSidebar: VaSidebar,
  VaSidebarItem: VaSidebarItem,
  VaSidebarItemContent: VaSidebarItemContent,
  VaSidebarItemTitle: VaSidebarItemTitle,
  VaSlider: VaSlider,
  VaSpacer: script$1,
  VaSwitch: VaSwitch,
  VaTabs: VaTabs,
  VaTab: VaTab,
  VaTimeInput: VaTimeInput,
  VaTimeline: VaTimeline,
  VaTimelineItem: VaTimelineItem,
  VaTimePicker: VaTimePicker,
  VaTreeCategory: VaTreeCategory,
  VaTreeNode: VaTreeNode,
  VaTreeRoot: VaTreeView,
  VaTreeView: VaTreeView
});

export { vuesticComponentMap as v };
